### PR TITLE
OV-16: feat-integrate-azure-ai

### PR DESCRIPTION
**Pull request by JPjok3r**
This is what I had, so I missunderstood every thing, I thought that we needed to render the avatar video, in real time, and speak a certain text string, I've been strugling with `getAvatar()` method in the class azure in backend/src/common/services/azure, and `textToSpeech()` is on it's right track, returning the audio, I'll work on this right now to modify the `getAvatar `.

### CODE REVIEW

**Comment by JPjok3r** [comment]
Hello @nikita-remeslov, @sofiia-trokhymchuk , @anton-otroshchenko this is what I have now and from what I researched, azure does not provide and endpoint to get avater images neither from the SDK, what implementation do is get all available voices names to convert text to speech.

File: backend/src/bundles/azure/azure.controller.ts
Comment (nikita-remeslov) [inline backend/src/bundles/azure/azure.controller.ts:1]: AZURE is just the SDK that we use! If we change SDK we need to change the bundle name?

File: backend/src/bundles/azure/azure.controller.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+     *       200:
+     *         description: Successful response
+     *         content:
+     *           application/json:
+     *             schema:
+     *               type: array
+     *               items:
+     *                 $ref: '#/components/schemas/VoiceModel'
+     */
+    private async getVoices(): Promise<ApiHandlerResponse> {
```
Comment (nikita-remeslov) [inline backend/src/bundles/azure/azure.controller.ts:56]: we need to load avatar info as well, not just the voice

File: backend/src/bundles/azure/azure.controller.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+     *     responses:
+     *       200:
+     *         description: Successful response
+     *         content:
+     *           audio/mpeg:
+     *             schema:
+     *               type: string
+     *               format: binary
+     */
+    private async generateSpeech(
```
Comment (nikita-remeslov) [inline backend/src/bundles/azure/azure.controller.ts:85]: we need return URL with uploaded audio

File: backend/src/common/services/azure/azure-ai.service.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
+import axios from 'axios';
```
Comment (nikita-remeslov) [inline backend/src/common/services/azure/azure-ai.service.ts:1]: we have http service to ferch data

File: backend/src/common/services/azure/azure-ai.service.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+import { type BaseConfig } from '~/common/config/base-config.package.js';
+import { HttpError } from '~/common/http/http.js';
+
+import { type AzureGetAvatarResponseDto } from './types/types.js';
+
+class AzureAIService {
```
Comment (nikita-remeslov) [inline backend/src/common/services/azure/azure-ai.service.ts:8]: Looks like we need

- voices
- avatar info
- text to speech
- get full rendered video

File: backend/package.json [hunk 5]
```diff
@@ -35,13 +35,15 @@
...
         "@fastify/swagger-ui": "4.0.1",
+        "axios": "1.7.5",
         "bcrypt": "5.1.1",
         "convict": "6.2.4",
         "dotenv": "16.4.5",
         "fastify": "4.28.1",
         "fastify-plugin": "4.5.1",
         "jose": "5.7.0",
         "knex": "3.1.0",
+        "microsoft-cognitiveservices-speech-sdk": "1.40.0",
```
Comment (nikita-remeslov) [inline backend/package.json:46]: looks like you even don't use this sdk

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]
Did you read this doc? - https://learn.microsoft.com/en-us/azure/ai-services/speech-service/get-started-text-to-speech?tabs=macos%2Cterminal&pivots=programming-language-javascript

File: backend/package.json (hunk 5 above)
Comment (JPjok3r) [reply backend/package.json:47]: I'm using it now for text to speech

File: backend/package.json [hunk 6]
```diff
@@ -35,13 +35,15 @@
         "@fastify/static": "7.0.4",
         "@fastify/swagger": "8.15.0",
         "@fastify/swagger-ui": "4.0.1",
+        "axios": "1.7.5",
```
[the remark this answers is not in the document]
Comment (JPjok3r) [reply backend/package.json]: axios gone

File: backend/src/common/services/azure/azure-ai.service.ts (hunk 4 above)
Comment (JPjok3r) [reply backend/src/common/services/azure/azure-ai.service.ts:18]: as for now it is not merged yet avatar info, I return voices filtered by language EN, the URL from result of text to speech, and I added two mwthods to create a batchSynthesisVideo, and get the url of that video, we should receive SSML script like here: https://learn.microsoft.com/en-us/azure/ai-services/speech-service/text-to-speech-avatar/avatar-gestures-with-ssml#supported-prebuilt-avatar-characters-styles-and-gestures, where we can add gestures

File: backend/src/bundles/avatar/avatar.controller.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    private azureAIService: AvatarService;
+
+    public constructor(logger: Logger, azureAIService: AvatarService) {
+        super(logger, ApiPath.AZURE_AI);
+
+        this.azureAIService = azureAIService;
+
+        this.addRoute({
+            path: AvatarApiPath.AVATAR_VOICES,
+            method: 'GET',
```
Comment (sofiia-trokhymchuk) [inline backend/src/bundles/avatar/avatar.controller.ts:24]: Use HttpMethod enum everywhere

File: backend/src/bundles/avatar/avatar.service.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        this.azureAIService = azureAIService;
+    }
+
+    public async getVoices(): Promise<GetAvatarVoicesResponseDto[]> {
+        try {
+            return await this.azureAIService.getAvatarVoices();
+        } catch (error) {
+            throw new HttpError({
+                message: 'Failed to fetch voices',
+                status: 500,
```
Comment (sofiia-trokhymchuk) [inline backend/src/bundles/avatar/avatar.service.ts:19]: 1. Use HttpCode enum values
2. Internal Server Errors are handled automatically in [BaseServerApp](https://github.com/BinaryStudioAcademy/bsa-2024-outreachvids/blob/next/backend/src/common/server-application/base-server-app.ts). It is better to add custom messages enum. Use unified approach all over the project

File: backend/src/common/services/azure/azure-ai.service.ts [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+import { fileService } from '../services.js';
+import { type GetAvatarVoicesResponseDto } from './types/types.js';
+
+interface batchSynthesesAvatar {
+    avatarName: string;
+    avatarStyle: string;
+    textSSML: string;
+    voice: string;
+}
```
Comment (sofiia-trokhymchuk) [inline backend/src/common/services/azure/azure-ai.service.ts:16]: Use types, not interfaces. Also, types and interfaces should start from upper-case

File: backend/src/common/services/azure/azure-ai.service.ts [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+            .filter(
+                (data: GetAvatarVoicesResponseDto) => data.Locale === 'en-US',
+            )
+            .map((data: GetAvatarVoicesResponseDto) => ({
+                DisplayName: data.DisplayName,
+                ShortName: data.ShortName,
+                Locale: data.Locale,
+                Gender: data.Gender,
+                LocaleName: data.LocaleName,
+                StyleList: data.StyleList,
```
Comment (sofiia-trokhymchuk) [inline backend/src/common/services/azure/azure-ai.service.ts:58]: Returned payload should be in camelCase

File: backend/src/common/services/azure/azure-ai.service.ts [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+                        );
+                    }
+                    speechSynthesizer.close();
+                },
+                (error) => {
+                    speechSynthesizer.close();
+                    reject(error);
+                },
+            );
+        });
```
Comment (sofiia-trokhymchuk) [inline backend/src/common/services/azure/azure-ai.service.ts:104]: Don't mix async-await style with promises

**Review by sofiia-trokhymchuk — CHANGES_REQUESTED** [review]

**Comment by sofiia-trokhymchuk** [comment]
Also, separate logical units of code blocks with an empty line. It is difficult to read the code when it looks like a solid canvas

**Comment by o-nedashkivska** [comment]
You should update .env.example file with new env vars

File: backend/src/bundles/avatar/avatar.controller.ts (hunk 7 above)
Comment (JPjok3r) [reply backend/src/bundles/avatar/avatar.controller.ts]: HttpMethod applied everywhere

File: backend/src/bundles/avatar/avatar.service.ts (hunk 8 above)
Comment (JPjok3r) [reply backend/src/bundles/avatar/avatar.service.ts]: deleted internal server error and changed using HttpCode. Also with custom message enum

File: backend/src/common/services/azure/azure-ai.service.ts (hunk 9 above)
Comment (JPjok3r) [reply backend/src/common/services/azure/azure-ai.service.ts]: Using type now, also checked upper-case starting

File: backend/src/common/services/azure/azure-ai.service.ts (hunk 11 above)
Comment (JPjok3r) [reply backend/src/common/services/azure/azure-ai.service.ts]: Promise moved to a privet method to be called and in public textToSpeech method only async-await

**Pull request closed without merging**